File: mobile/android/app/src/main/java/fyi/dayle/data/MissionRepository.kt
package fyi.dayle.data

class MissionRepository(
    private val database: DayleDatabase, private val service: MissionService
) {

    fun getTodayMission(): Mission? = database.getMissionForDate(DateUtil.today())

    fun setCompleted(mission: Mission, completed: Boolean) {
        database.setCompleted(mission.date, completed)
    }

    fun fetchAndStoreTodayMission(): Mission {
        val previous = database.getRecentMissions(RECENT_LIMIT).map { it.text }
        val text = service.fetchMission(previous)
        return database.insertMission(DateUtil.today(), text)
    }

    companion object {
        private const val RECENT_LIMIT = 10
    }
}

File: app/lib/services/mission_repository.dart
import "../models/mission.dart";
import "../utils/date_util.dart";
import "database_service.dart";
import "mission_service.dart";

class MissionRepository {
  static const _recentLimit = 10;

  final DayleDatabase _database;
  final MissionService _service;

  MissionRepository(this._database, this._service);

  Future<Mission?> getTodayMission() async {
    return _database.getMissionForDate(DateUtil.today());
  }

  Future<void> setCompleted(Mission mission, bool completed) async {
    await _database.setCompleted(mission.date, completed);
  }

  Future<Mission> fetchAndStoreTodayMission() async {
    final recent = await _database.getRecentMissions(_recentLimit);
    final previous = recent.map((m) => m.text).toList();
    final text = await _service.fetchMission(previous);
    return _database.insertMission(DateUtil.today(), text);
  }
}
